feat: Add support for events

ssmucny opened the pull request:
<!-- ℹ お読みください / README
PRありがとうございます！ PRを作成する前に、コントリビューションガイドをご確認ください:
Thank you for your PR! Before creating a PR, please check the contribution guide:
https://github.com/misskey-dev/misskey/blob/develop/CONTRIBUTING.md
-->

## What
<!-- このPRで何をしたのか？ どう変わるのか？ -->
<!-- What did you do with this PR? How will it change things? -->
Adding support for events in Misskey. See #3217 

### New `Event` type
Implemented like `Polls` (a `Note` can have a child event in the database). The event data will show above the normal content of displayed notes if present.

The required values in a child event are `start` and `title`. The `end` and `metadata` are optional. The metadata is an Schema.org JSON object of type [Event](https://schema.org/Event). The event editor provides the ability for users to specify elements in the metadata, but the API endpoint can support more properties than the UI.

### New API for searching Events
New endpoint created at `/notes/events/search` that queries events. Similar parameters to other searches with changes for Event specific values. The major addition are new filter options. These are objects that filter events based on metadata. Each object is treated as a separate AND condition. Each object contains a key array that selects the metadata property to filter on (ex:  `["location", "postalCode"]` selects location.postalCode in the Event metadata. Each object also has an array of string values that are considered as OR conditions for their respective metadata property. So if the prior location.postaCode selection wanted to select all events that had postal code 44113 and 44114 then they would use ["44113","44114"]. A value of `null` indicates that a missing value in the event is matched.

The individual values are interpreted as case insensitive regex values, so the user could run more complicated matches themselves; however, because regex is accepted, there is a timeout placed on the query (250 ms) to prevent long running queries from hurting performance (advice from [Postgres docs](https://www.postgresql.org/docs/current/functions-matching.html)). There can be multiple filters using the same metadata key, and these will have the effect of an AND condition on the same metadata value with two different match clauses.

The original issue has a proposal for an alternative API using a query string parameters in the URL. This PR does not implement this, but it could be implemented in a separate PR.

### Event display component
A new Vue component to display events in notes was created. It formats the note information at the top of the note. Other note stuff is still accessible.
![image](https://github.com/misskey-dev/misskey/assets/17689855/1ea7a018-d6aa-4f52-9365-a44d33ca6d02)

### User Events page
Added a tab in the user details page to show their events. The events can be displayed in reverse chronological order (normal display order) or with events shown in date order (starting with the events that are most imminent). If showing imminent events, then past events are not shown.
![image](https://github.com/misskey-dev/misskey/assets/17689855/a0195d9e-fe47-4983-8dde-fc9021a2c20e)

### Search Events page
Added a tab in the search page to search events. There is a custom form with parameters relevant to searching events. This includes a query string that searches the `title` and `text` of events, start/end date filters, and a option to change the sort order (like with the User Events page). Currently form elements to specify user filters and metadata filters are not included (using the API directly is the only way to filter with them). These should be added in a later PR.
![image](https://github.com/misskey-dev/misskey/assets/17689855/279f97f6-1898-4d29-83df-b1d72b0a0b6c)

### Event editor
Added an option in the note creation dialogue to make a note an event. When selected, extra forms appear where the user can fill in details. The title and start date/time are required while the rest are optional. A toggle is present that hides most of the advanced options unless needed.
![image](https://github.com/misskey-dev/misskey/assets/17689855/9727619a-3ee2-4c68-a25c-b4f8d3582ed5)

## Why
<!-- なぜそうするのか？ どういう意図なのか？ 何が困っているのか？ -->
<!-- Why do you do it? What are your intentions? What is the problem? -->
Enable users to create events to share and track activities that occur at a particular time.

## Additional info
<!-- テスト観点など -->
<!-- Test perspective, etc -->
### Implementation progress:
- [x] new Event table that is One-to-One with Note
- [x] services/queries to manage events
- [x] update API endpoints to include events (notes/[create|delete|show])
- [x] Document new API changes
- [x] new API endpoints to search events (notes/events/search)
- [x] UI elements to display events
- [x] UI elements to create events
- [ ] Event export (iCal) (maybe)
- [x] Event federation ([Activity Pub Event](https://docs.joinmobilizon.org/contribute/activity_pub/))
- [ ] RSVP (maybe)
- [ ] Edit Note/Event (and federate) (https://github.com/misskey-dev/misskey/issues/8362) (maybe)

## Checklist
- [x] Read the [contribution guide](https://github.com/misskey-dev/misskey/blob/develop/CONTRIBUTING.md)
- [x] Test working in a local environment
- [x] (If needed) Update CHANGELOG.md
- [x] (If possible) Add tests

ssmucny committed:
Add event source code to models

ssmucny committed:
entity typo

ssmucny committed:
Event migration

ssmucny commented on packages/backend/migration/1681429921400-Event.js:
```diff
@@ -0,0 +1,95 @@
+    name = 'Event1681429921400'
+
+    async up(queryRunner) {
+        await queryRunner.query(`DROP INDEX "public"."IDX_2cd3b2a6b4cf0b910b260afe08"`);
```
The migration generated appears to have a bunch of extraneous modifications. Should these be removed so that only the new Event related modifications are included?

remador commented:
Coruja N has sent you an email via Gmail confidential mode:

[image: Gmail logo]Re: [misskey-dev/misskey] feat: Add support for events (PR #10628) <https://confidential-mail.google.com/msg/AA2T6PaW7I2FJEym0HNMq5gzqIO4-AZpMaDejpC-m01KuWdkycDNck4IspzdxweAyikFdtEN-W40xP4oj9fdpTgxdOmtv09L7Os6eZwSHEkWR8vtZRi3SZuw1QbmRtiKaj4H6C-upU2ghEGfzAbrT_6eZKqC0k_ElQPHwIJoHm78dv2zFoz8TcjJLUHiyfQ8w7YW7iWjpyA0SZM=>

This message was sent on Apr 13, 2023 at 7:43:56 PM PDT
You can open it by clicking the link below. This link will only work for ***@***.***

View the email <https://confidential-mail.google.com/msg/AA2T6PaW7I2FJEym0HNMq5gzqIO4-AZpMaDejpC-m01KuWdkycDNck4IspzdxweAyikFdtEN-W40xP4oj9fdpTgxdOmtv09L7Os6eZwSHEkWR8vtZRi3SZuw1QbmRtiKaj4H6C-upU2ghEGfzAbrT_6eZKqC0k_ElQPHwIJoHm78dv2zFoz8TcjJLUHiyfQ8w7YW7iWjpyA0SZM=>

Gmail confidential mode gives you more control over the messages you send. The sender may have chosen to set an expiration time, disable printing or forwarding, or track access to this message. Learn more <https://support.google.com/mail/answer/7674059>
Gmail: Email by Google
Use is subject to the Google Privacy Policy <https://myaccount.google.com/privacypolicy?hl=en>
Google LLC, 1600 Amphitheatre Parkway, Mountain View, CA 94043, USA
You have received this message because someone sent you an email via Gmail confidential mode.
[image: Google logo]

remador commented:
Your Google verification code is 487736

remador commented:
您的 Google 验证码是 874259

ssmucny committed:
Changed DB to have events behave like polls

ssmucny committed:
add new migration

ssmucny committed:
denormalize Events/Notes relation

ssmucny committed:
Add Event to create api

ssmucny committed:
Add event to repositories (getters)

ssmucny committed:
Added notes/events/search endpoint logic

ssmucny committed:
Add simple display of event to note in UI

ssmucny committed:
Add events tab to user details page

ssmucny committed:
Change displayed date/time format for events

ssmucny committed:
Add form to create events in UI

ssmucny committed:
Add ability to GET event data from federated AP

ssmucny committed:
Add event to AP publishing

ssmucny committed:
Add basic event search page

ssmucny committed:
Add query string to event search

ssmucny committed:
Add date range to event search page

ssmucny committed:
Update API types

ssmucny committed:
bug fixes and localization:
- add localization strings
- fixed duplicates in search (use offset when sorting by startDate)

ssmucny committed:
update CHANGELOG

ssmucny committed:
Merge branch 'develop' into ssmucny-events

ssmucny committed:
Bug fix in extractEventFromNote with minimum Note

ssmucny committed:
Update misskey-js API md file

ssmucny committed:
lint fix

ssmucny committed:
Change event search to use similar to clause
- enable more customized searches (and/or nesting)

ssmucny committed:
update API docs

ssmucny committed:
Change event search
- case insensitive regex
- use object type for filters param
- update API docs

ssmucny committed:
Merge branch 'develop' into ssmucny-events

ssmucny commented on packages/backend/src/models/entities/Event.ts:
```diff
@@ -0,0 +1,79 @@
+		default: {},
+		comment: 'metadata mapping for event with more user configurable optional information',
+	})
+	public metadata: Record<string, string>;
```
Would it be better to have this be an arbitrary object type to accept more customized event schema? For example an address field (https://schema.org/PostalAddress). The drawback with this is displaying/searching a more complicated object would be more difficult than just text. A GraphQL API might be better suited for this, but we don't have one.

acid-chicken replied:
It would be better to fix it to a schema like PostalAddress. Accepting arbitrary fields would probably result in too many notational gaps and would be practically unusable. Instead, it is more user-friendly for the software to indicate the fields that should be set.

acid-chicken committed:
chore: tweak MkEvent

ssmucny committed:
Change events to use schema.org/Event for metadata

ssmucny committed:
Merge branch 'ssmucny-events' of https://github.com/ssmucny/misskey into ssmucny-events

ssmucny committed:
Add Event schema to metadata
- add types from https://schema.org/Event
- updated MkEvent to display new fields
- minor refactoring

ssmucny committed:
update APEventService with new metadata schema

ssmucny committed:
Updating event editor to handle new fields

ssmucny committed:
Merge branch 'develop' into ssmucny-events

ssmucny committed:
update event editor fields

ssmucny committed:
Merge branch 'ssmucny-events' of https://github.com/ssmucny/misskey into ssmucny-events

ssmucny committed:
Fixes to event editor

ssmucny committed:
- add i18n
- upate UI storybook

ssmucny committed:
- add links to event display
- add fold expansion for advanced event options
- index update

ssmucny committed:
Merge branch 'develop' into ssmucny-events

ssmucny committed:
fix event search

ssmucny committed:
UI fixes

ssmucny committed:
- refactor search page
- lint fixes
- UI bug fix

ssmucny committed:
update api types

ssmucny committed:
lint fixes

ssmucny committed:
update event types for federation

ssmucny commented:
This PR has all the major features added and is ready for reviews.

ThatOneCalculator commented on packages/frontend/src/components/MkEventEditor.vue:
```diff
@@ -0,0 +1,272 @@
+	>.caution {
+		margin: 0 0 8px 0;
+		font-size: 0.8em;
+		color: #f00;
```
This should likely `var(--warn)`

ssmucny replied:
These styles aren't even needed. Just dead code.

ssmucny committed:
Clean up styles

ssmucny committed:
Merge branch 'develop' into ssmucny-events

ssmucny committed:
Merge branch 'develop' into ssmucny-events

tamaina commented:
I honestly can't think of a situation where I would use it, but if there was a webcal (iCal) link, I might want to use it.

ThatOneCalculator commented:
I'm currently adapting this PR, and using [ics](https://www.npmjs.com/package/ics) for iCal generation

tamaina commented:
Need to get @syuilo to review the design of the content in the note.

tamaina committed:
Merge branch 'develop' into ssmucny-events

syuilo commented:
見る

ssmucny committed:
Merge branch 'develop' into ssmucny-events

syuilo commented:
見てる

syuilo committed:
Merge branch 'develop' into ssmucny-events

syuilo commented:
半分くらい見た
これってMastodonとか他のソフトウェアは対応してるのかな

acid-chicken commented:
Event 実装といえば Friendica が挙がりそうだけど連合できるかしら

ssmucny committed:
Merge branch 'develop' into ssmucny-events

ssmucny commented:
Friendica and Mobilizon both have ActivityPub Events. They have some extensions to the base object.

Friendica: https://activitystrea.ms/head/activity-schema.html#event, https://github.com/friendica/friendica/wiki/ActivityStreams#activity_obj_event
Mobilizon: https://docs.joinmobilizon.org/contribute/activity_pub/#event_1
ActivityPub: https://www.w3.org/TR/activitystreams-vocabulary/#dfn-event

Mastodon does not have Events

syuilo commented:
概ね良さそう
ただEventという名前がプログラミングにおいては一般的すぎるから変えた方が良いかもと思った

syuilo commented:
例えばEventが必ずNoteと紐づけられるのであればNoteEventとか？

syuilo commented:
あとはEventAnnoucementとか思いついた

tamaina commented:
Eventから離れたほうが良いと思う

Scheduleとか…Programとか…あれ、両方コンピュータで使われがちだ…

tamaina commented:
Sessionもまずい

syuilo commented:
Off topic:
突き詰めて考えると全てのドメイン固有概念にはVueコンポーネントでやってるみたいにMkを接頭辞として付けるみたいなの思いついた

ssmucny commented:
> 例えばEventが必ずNoteと紐づけられるのであればNoteEventとか？

I like this idea. 'Event' is already used in the repository to refer to other objects, so changing the name to 'NoteEvent' internally would help to disambiguate.

syuilo commented:
> Off topic: 突き詰めて考えると全てのドメイン固有概念にはVueコンポーネントでやってるみたいにMkを接頭辞として付けるみたいなの思いついた

https://github.com/misskey-dev/misskey/issues/11496

syuilo committed:
Merge branch 'develop' into pr/10628

syuilo committed:
Update Event.ts

ssmucny committed:
Merge branch 'develop' into ssmucny-events

syuilo committed:
Merge branch 'develop' into pr/10628

syuilo committed:
Merge branch 'ssmucny-events' of https://github.com/ssmucny/misskey into pr/10628

syuilo committed:
tweak

syuilo committed:
Update index.d.ts

syuilo commented on packages/backend/src/models/entities/Event.ts:
```diff
@@ -0,0 +1,119 @@
+	})
+	public title: string;
+
+	@Column('jsonb', {
+		default: {
+			'@context': 'https://schema.org/',
+			'@type': 'Event',
+		},
+		comment: 'metadata object describing the event. Follows https://schema.org/Event',
+	})
+	public metadata: EventSchema;
```
スキーマが予め決まっているデータだからJSONBでまとめて持つ必要はなさそう

ssmucny replied:
Schema.org has a defined schema, but not all platforms implement the same properties from it. Keeping it in JSONB ensures compatibility with other sources and flexibility if we want to support other properties later. Some common/required properties like 'start time' and 'end time' that will be queried/ordered often are included as their own SQL fields with indexes even if that adds some redundancy with the JSONB object.
